File: src/weaponsprocurement/core/StockPurchasePlan.java
package weaponsprocurement.core;

import java.util.ArrayList;
import java.util.List;

final class StockPurchasePlan {
    final List<StockPurchaseLine> lines;
    final int totalQuantity;
    final long totalCost;
    final float totalSpace;

    private StockPurchasePlan(List<StockPurchaseLine> lines, int totalQuantity, long totalCost, float totalSpace) {
        this.lines = lines;
        this.totalQuantity = totalQuantity;
        this.totalCost = totalCost;
        this.totalSpace = totalSpace;
    }

    static StockPurchasePlan build(List<StockPurchaseSource> sources, int requestedQuantity) {
        int remaining = requestedQuantity;
        int totalQuantity = 0;
        long totalCost = 0L;
        float totalSpace = 0f;
        List<StockPurchaseLine> lines = new ArrayList<StockPurchaseLine>();
        for (StockPurchaseSource source : sources) {
            if (remaining <= 0) {
                break;
            }
            int quantity = Math.min(remaining, source.available);
            if (quantity <= 0) {
                continue;
            }
            lines.add(new StockPurchaseLine(source, quantity));
            remaining -= quantity;
            totalQuantity += quantity;
            totalCost = TradeMoney.safeAdd(totalCost, TradeMoney.lineTotal(source.unitPrice, quantity));
            totalSpace += source.unitCargoSpace * quantity;
        }
        return new StockPurchasePlan(lines, totalQuantity, totalCost, totalSpace);
    }
}

File: src/main/kotlin/weaponsprocurement/core/StockPurchasePlan.kt
package weaponsprocurement.core

import java.util.ArrayList

class StockPurchasePlan private constructor(
    @JvmField val lines: List<StockPurchaseLine>,
    @JvmField val totalQuantity: Int,
    @JvmField val totalCost: Long,
    @JvmField val totalSpace: Float,
) {
    companion object {
        @JvmStatic
        fun build(sources: List<StockPurchaseSource>, requestedQuantity: Int): StockPurchasePlan {
            var remaining = requestedQuantity
            var totalQuantity = 0
            var totalCost = 0L
            var totalSpace = 0f
            val lines = ArrayList<StockPurchaseLine>()
            for (source in sources) {
                if (remaining <= 0) break
                val quantity = remaining.coerceAtMost(source.available)
                if (quantity <= 0) continue
                lines.add(StockPurchaseLine(source, quantity))
                remaining -= quantity
                totalQuantity += quantity
                totalCost = TradeMoney.safeAdd(totalCost, TradeMoney.lineTotal(source.unitPrice, quantity))
                totalSpace += source.unitCargoSpace * quantity
            }
            return StockPurchasePlan(lines, totalQuantity, totalCost, totalSpace)
        }
    }
}
